source/routing/journey_pattern_container.h

```diff
@@ -0,0 +1,191 @@
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+Stay tuned using
+twitter @navitia
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+
+#pragma once
+
+#include "raptor_utils.h"
+#include <boost/optional.hpp>
+
+namespace navitia { namespace type {
+
+struct PT_Data;
+struct DiscreteVehicleJourney;
+struct FrequencyVehicleJourney;
+struct StopTime;
+
+}}// namespace navitia::type
+
+namespace navitia { namespace routing {
+
+// TODO: constructor private with JourneyPatternContainer friend?
+struct JourneyPatternPoint {
+    JpIdx jp_idx;
+    SpIdx sp_idx;
+    uint16_t order;
+    bool operator==(const JourneyPatternPoint& other) const {
+        return sp_idx == other.sp_idx && order == other.order;
+    }
+};
+std::ostream& operator<<(std::ostream&, const JourneyPatternPoint&);
+
+
+// TODO: constructor private with JourneyPatternContainer friend?
+struct JourneyPattern {
+    std::vector<JppIdx> jpps;
+    std::vector<const type::DiscreteVehicleJourney*> discrete_vjs;
+    std::vector<const type::FrequencyVehicleJourney*> freq_vjs;
+    RouteIdx route_idx;
+    PhyModeIdx phy_mode_idx;
+
+    bool operator==(const JourneyPattern& other) const {
+        return jpps == other.jpps
+            && discrete_vjs == other.discrete_vjs
+            && freq_vjs == other.freq_vjs
+            && route_idx == other.route_idx;
+    }
+    template<typename VJ> std::vector<const VJ*>& get_vjs();
+    template<typename F> void for_each_vehicle_journey(const F f) const {
+        for (const auto& vj: discrete_vjs) { if (! f(*vj)) { return; } }
+        for (const auto& vj: freq_vjs) { if (! f(*vj)) { return; } }
+    }
+};
+std::ostream& operator<<(std::ostream&, const JourneyPattern&);
+
+
+struct JourneyPatternContainer {
+    using Jps = std::vector<JourneyPattern>;
+    using Jpps = std::vector<JourneyPatternPoint>;
+    using JpIterator = IdxMapIterator<JpIdx, Jps::const_iterator>;
+    using JppIterator = IdxMapIterator<JppIdx, Jpps::const_iterator>;
+    using JpRange = boost::iterator_range<JpIterator>;
+    using JppRange = boost::iterator_range<JppIterator>;
+
+    void load(const navitia::type::PT_Data&);
+    size_t nb_jps() const { return jps.size(); }
+    size_t nb_jpps() const { return jpps.size(); }
+    const JourneyPattern& get(const JpIdx& idx) const {
+        assert(idx.val < jps.size());
+        return jps[idx.val];
+    }
+    const JourneyPatternPoint& get(const JppIdx& idx) const {
+        assert(idx.val < jpps.size());
+        return jpps[idx.val];
```

Comment:
you can actually call `std::vector<T>::at()` which throws an exception when index is out of range, if i'm not wrong